src/applets/budgie-menu/OverlayMenus.vala

```diff
@@ -0,0 +1,313 @@
+		this.power_menu_items.add(logout_menu);
+		this.power_menu_items.add(suspend_menu);
+		this.power_menu_items.add(hibernate_menu);
+		this.power_menu_items.add(reboot_menu);
+		this.power_menu_items.add(shutdown_menu);
+
+		this.stack.add_named(this.folder_items, "xdg");
+		this.stack.add_named(this.power_menu_items, "power");
+		this.add(this.stack);
+
+		this.setup_menu_events();
+	}
+
+	private bool filter_list_box_item(ListBoxRow row) {
+		MenuItem menu_item = (MenuItem) row.get_child();
+		UserDirectory xdg_dir = menu_item.get_data("user-directory");
+		return (this.items_to_show.index(xdg_dir) != -1);
+	}
+
+	private void handle_xdg_dir_clicked(Button item) {
+		UserDirectory xdg_dir = item.get_data("user-directory");
+		unowned string? path = Environment.get_user_special_dir(xdg_dir);
+
+		if (path == null) {
+			return;
+		}
+
+		Gdk.AppLaunchContext launch_context = (Display.get_default()).get_app_launch_context(); // Get the app launch context for the default display
+		launch_context.set_screen(Screen.get_default()); // Set the screen
+		launch_context.set_timestamp(CURRENT_TIME);
+
+		DesktopAppInfo? appinfo = (DesktopAppInfo) AppInfo.get_default_for_type("inode/directory", true); // Ensure we using something which can handle inode/directory
+		List<string> files = new List<string>();
+		files.append("file://"+path);
+
+		try {
+			appinfo.launch_uris(files, launch_context);
+		} catch (Error e) {
+			warning("Failed to open %s: %s", path, e.message);
+		}
+	}
+
+	/**
+	 * Set up all of the DBus bits to make all the items work.
+	 */
+	private async void setup_dbus() {
+		try {
+			this.logind_interface = yield Bus.get_proxy(BusType.SYSTEM, LOGIND_LOGIN, "/org/freedesktop/login1");
+		} catch (Error e) {
+			warning("Unable to connect to logind: %s", e.message);
+		}
+
+		try {
+			this.saver = yield Bus.get_proxy(BusType.SESSION, "org.gnome.ScreenSaver", "/org/gnome/ScreenSaver");
+		} catch (Error e) {
+#if HAVE_GNOME_SCREENSAVER
+			warning("Unable to connect to gnome-screensaver: %s", e.message);
+#else
+			warning("Unable to connect to budgie-screensaver: %s", e.message);
+#endif
+			return;
+		}
+
+		try {
+			this.session = yield Bus.get_proxy(BusType.SESSION, G_SESSION, "/org/gnome/SessionManager");
+		} catch (Error e) {
+			warning("Unable to connect to GNOME Session: %s", e.message);
+		}
+
+		try {
+			this.xdgtracker = yield Bus.get_proxy(BusType.SESSION, "org.buddiesofbudgie.XDGDirTracker", "/org/buddiesofbudgie/XDGDirTracker");
+			this.xdgtracker.xdg_dirs_exist.connect(this.handle_xdg_dirs_changed);
+			this.handle_xdg_dirs_changed(this.xdgtracker.get_dirs());
+		} catch (Error e) {
+			warning("Unable to connect to Budgie XDGDirTracker: %s", e.message);
+		}
+	}
+
+	private void handle_xdg_dirs_changed(UserDirectory[] dirs) {
+		this.items_to_show = new List<UserDirectory>(); // Remove all items
+
+		for (var i = 0; i < dirs.length; i++) { // For each directory
+			this.items_to_show.append(dirs[i]);
```

Comment:
Instead of sorting each time, could we do `insert_sorted()`? I suppose it might be a case of six of one, and half a dozen of the other.

Reply:
We aren't sorting the items_to_show as they do not need to be sorted. The items being sorted are folder_items and those do not get sorted during the handling of the incoming XDG dirs list from the DBus Server. What is being called after the new List is being created is the filtering invalidation, since we are re-checking if an item should be shown / hidden (e.g. Desktop folder gets deleted -> hide item, Music folder gets created -> show item).